contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,333 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.20;
+
+import {Hooks} from "@uniswap/v4-core/src/libraries/Hooks.sol";
+import {TickMath} from "@uniswap/v4-core/src/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/src/interfaces/IHooks.sol";
+import {ILockCallback} from "@uniswap/v4-core/src/interfaces/callback/ILockCallback.sol";
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {Currency} from "@uniswap/v4-core/src/types/Currency.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {PoolIdLibrary} from "@uniswap/v4-core/src/types/PoolId.sol";
+import {IQuoter} from "../interfaces/IQuoter.sol";
+import {PoolTicksCounter} from "../libraries/PoolTicksCounter.sol";
+import {PathKey, PathKeyLib} from "../libraries/PathKey.sol";
+
+contract Quoter is IQuoter, ILockCallback {
+    using Hooks for IHooks;
+    using PoolIdLibrary for PoolKey;
+    using PathKeyLib for PathKey;
+
+    /// @dev cache used to check a safety condition in exact output swaps.
+    uint256 private amountOutCached;
+
+    // v4 Singleton contract
+    IPoolManager public immutable manager;
+
+    /// @dev custom error function selector length
+    uint256 internal constant MINIMUM_CUSTOM_ERROR_LENGTH = 4;
+
+    /// @dev function selector + length of bytes as uint256 + min length of revert reason padded to multiple of 32 bytes
+    uint256 internal constant MINIMUM_REASON_LENGTH = 68;
+
+    /// @dev min valid reason is 3-words long
+    /// @dev int128[2] + sqrtPriceX96After padded to 32bytes + intializeTicksLoaded padded to 32bytes
+    uint256 internal constant MINIMUM_VALID_RESPONSE_LENGTH = 96;
+
+    /// @dev Only this address may call this function
+    modifier selfOnly() {
+        if (msg.sender != address(this)) revert NotSelf();
+        _;
+    }
+
+    constructor(address _poolManager) {
+        manager = IPoolManager(_poolManager);
+    }
+
+    /// @inheritdoc IQuoter
+    function quoteExactInputSingle(QuoteExactSingleParams memory params)
+        public
+        override
+        returns (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksLoaded)
+    {
+        try manager.lock(address(this), abi.encodeWithSelector(this._quoteExactInputSingle.selector, params)) {}
+        catch (bytes memory reason) {
+            return _handleRevertSingle(reason);
+        }
+    }
+
+    /// @inheritdoc IQuoter
+    function quoteExactInput(QuoteExactParams memory params)
+        external
```

Comment:
this should be public right? if we are doing the batched self call

Reply:
I think you are right but would rather change in the separate PR that uses batched self call